### ISSUE TITLE
Remove Unnecessary Macros From Code

### ISSUE THREAD

**Issue by gatoWololo** [boltlabs-inc/tss-ecdsa#152]
This issue is not super high priority. There are currently a few small macros used throughout the code. It would be great to replace these with equivalent, more readable, functions.

These should be fairly straightforward to change. The macros that wrap error handling should probably be made into methods on our error enum. On `errors.rs`:
- [ ] `serialize!`
- [ ] `deserialize!`
- [ ] `verify_err!`
- [ ] `bail!`
- [ ] `bail_context!`

There are two additional macros which seem extra magical and may be more work than the ones above. On `src/participant.rs`:
- [ ] `run_only_once!`
- [ ] `run_only_once_per_tag!`

**Comment by marsella** [comment]
I think this is a reasonable desire, but I'm not sure it makes sense as a single issue. Several of these are covered by other tickets (`bail` and `bail_context` by #127 , `verify_err` by #155).

For `serialize` and `deserialize`, there are at least three different cases where they're used, and each one needs fairly significant thinking. While I agree that the macros aren't ideal, I don't think it's worth it to do the grunt work of removing them independently / in advance of addressing these larger issues.
1. We've talked offline about updating the storage design to store rust types instead of serialized data. I think that refactor will remove most calls to those two macros, but there aren't explicit tickets about it yet.
2. Messages currently hold serialized bytes, which are supposed to be verified before they're deserialized. This requires a lot of manual calls to verify functions which is prone to misuse (indeed, in most places we don't verify received messages) and should be addressed in #32 (and more broadly, #6).
2. `serialize!` is also used to generate transcripts. This is a case where we actually do just want to serialize the bytes and squash them in, but if we're going to use it the same way every time, I'd be open to refactoring that API a bit. If you'd like, I could make another issue in #149 about updating the transcript API to call `serialize` internally, which we'll probably have to do in order to add all the other context anyway.

For the `run_only_once` and `run_only_once_per_tag` macros, at first glance these seem related to #160; we might be able to remove them by adding a `status` flag to storage that tracks not just done-ness but also protocol stage. As you note, that seems like a large enough project that it should probably be its own issue.

Thoughts?

**Comment by gatoWololo** [comment]
> I think this is a reasonable desire, but I'm not sure it makes sense as a single issue

Agreed. This issue was written more as a record and less as a ready-to-work-on ticket :sweat_smile: 

> While I agree that the macros aren't ideal, I don't think it's worth it to do the grunt work of removing them independently / in advance of addressing these larger issues.

Also agree!

It is pretty unfamiliar with how the parts of the code you listed work. So I defer to you for the best way to tackle them. Everything you listed sounds very reasonable though!

> I could make another issue in #149 about updating the transcript API to call serialize internally, which we'll probably have to do in order to add all the other context anyway.

That would be great if we could do that.

Breaking this down into smaller issue, keeping in mind the ordering they should be worked on, sounds like a great approach to me.
